ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/SszSuperNode.java

```diff
@@ -0,0 +1,172 @@
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static tech.pegasys.teku.ssz.backing.tree.GIndexUtil.gIdxCompare;
+import static tech.pegasys.teku.ssz.backing.tree.GIndexUtil.gIdxGetChildIndex;
+import static tech.pegasys.teku.ssz.backing.tree.GIndexUtil.gIdxGetRelativeGIndex;
+import static tech.pegasys.teku.ssz.backing.tree.GIndexUtil.gIdxIsSelf;
+
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.bytes.MutableBytes;
+import org.apache.tuweni.crypto.Hash;
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.teku.ssz.backing.tree.GIndexUtil.NodeRelation;
+import tech.pegasys.teku.ssz.backing.tree.SszNodeTemplate.Location;
+
+/**
+ * Stores consecutive elements of the same fixed size type as a single packed bytes of their leaves
+ * (this representation exactly matches SSZ representation of elements sequence).
+ *
+ * <p>This node represents a subtree of binary merkle tree for sequence (list or vector) of elements
+ * with maximum length of <code>2 ^ depth</code>. If the sequence has less than maximum elements
+ * then <code>ssz</code> bytes store only existing elements (what again matches SSZ representation
+ * of a list)
+ *
+ * <p>To address individual nodes inside elements and resolve their internal generalized indexes the
+ * node uses {@link SszNodeTemplate} which represents element type tree structure
+ *
+ * <p>This node favors memory efficiency over update performance and thus is the best choice for
+ * rarely updated and space consuming structures (e.g. Eth2 <code>BeaconState.validators</code>
+ * list)
+ */
+public class SszSuperNode implements TreeNode, LeafDataNode {
+  private static final TreeNode DEFAULT_NODE = LeafNode.EMPTY_LEAF;
+
+  private final int depth;
+  private final SszNodeTemplate elementTemplate;
+  private final Bytes ssz;
+  private final Supplier<Bytes32> hashTreeRoot = Suppliers.memoize(this::calcHashTreeRoot);
+
+  public SszSuperNode(int depth, SszNodeTemplate elementTemplate, Bytes ssz) {
+    this.depth = depth;
+    this.elementTemplate = elementTemplate;
+    this.ssz = ssz;
+    checkArgument(ssz.size() % elementTemplate.getSszLength() == 0);
+    checkArgument(getElementsCount() <= getMaxElements());
+  }
+
+  private int getMaxElements() {
+    return 1 << depth;
+  }
+
+  private int getElementsCount() {
+    return ssz.size() / elementTemplate.getSszLength();
+  }
+
+  @Override
+  public Bytes32 hashTreeRoot() {
+    return hashTreeRoot.get();
+  }
+
+  private Bytes32 calcHashTreeRoot() {
+    return hashTreeRoot(0, 0);
+  }
+
+  private Bytes32 hashTreeRoot(int curDepth, int offset) {
+    if (curDepth == depth) {
+      if (offset < ssz.size()) {
+        return elementTemplate.calculateHashTreeRoot(ssz, offset);
+      } else {
```

Comment:
This might make the code easier to read and make the code more defensive.
```suggestion
      } else if (offset == ssz.size()) {
        return DEFAULT_NODE.hashTreeRoot();
      } else {
        throw new IllegalStateException("Offset can not be greater than ssz length");
      }
```

Reply:
I've added the this check d6eec77 (as `assert` since this parameter is calculated internally and we probably wouldn't like to spend CPU in production for this) 
`offset` could be larger than `ssz.size()` cause you may have have ssz say just for 2 elements but the node capacity is say 256